libraries/tickv/src/lib.rs

```diff
@@ -0,0 +1,241 @@
+//!    .unwrap();
+//!
+//! // Add a key
+//! let value: [u8; 32] = [0x23; 32];
+//! tickv.append_key(&mut DefaultHasher::new(), b"ONE", &value).unwrap();
+//!
+//! // Get the same key back
+//! let mut buf: [u8; 32] = [0; 32];
+//! tickv.get_key(&mut DefaultHasher::new(), b"ONE", &mut buf).unwrap();
+//! ```
+//!
+//! You can then use the `get_key()` function to get the key back from flash.
+//!
+//! # Collisions
+//!
+//! TicKV will prevent a new key/value pair with a colliding hash of the key to be
+//! added. The collision will be reported to the user with the `KeyAlreadyExists`
+//! `ErroCode`.
+//!
+//! # Power loss protection
+//!
+//! TicKV ensures that in the event of a power loss, no stored data is lost or
+//! corrupted. The only data that can be lost in the event of a power loss is
+//! the data currently being written (if it hasn't been write to flash yet).
+//!
+//! If a power loss occurs after calling `append_key()` or `invalidate_key()`
+//! before it has completed then the operation probably did not complete and
+//! that data is lost.
+//!
+//! To help reduce this time to be as short as possible the `FlashController`
+//! is synchronous. Although flash writes can take a considerable amount of time
+//! and this will stall the application, this still seems like a good idea
+//! to avoid loosing data.
+//!
+//! # Security
+//!
+//! TicKV uses check sums to check data integrity. TicKV does not have any measures
+//! to prevent malicious manipulation or privacy. An attacker with access to the
+//! flash can change the values without being detected. An attacked with access
+//! to flash can also read all of the information. Any privacy, security or
+//! authentication measures need to be layered on top of TicKV.
+//!
+//! # Hash Function
+//!
+//! Any hash function that implements Rust's `core::hash::Hasher` trait can be used.
+//!
+//! The hash function ideally should generate uniform hashes and must not change during
+//! the lifetime of the filesystem.
+//!
+//! The Rust `core::hash::Hasher` implementation is a little strange. When the
+//! hash is calculated with the `finish()` function the internal state of the
+//! `Hasher` is not reset. This means that the check sum is generated with the
+//! following code and the key input becomes part of the check sum.
+//!
+//! ```rust,ignore
+//!         key.hash(hash_function);
+//!         let hash = hash_function.finish();
+//!
+//!         buf.hash(hash_function);
+//!         value.hash(hash_function);
+//!         let check_sum = hash_function.finish();
+//! ```
+//! ## Versions
+//!
+//! TicKV stores the version when adding objects to the flash storage.
+//!
+//! TicKV is currently version 0.
+//!
+//!  * Version 0
+//!    * Version 0 is a draft version. It should NOT be used for important data!
+//!      Version 0 maintains no backwards compatible support and could change at
+//!      any time.
+//!
+
+#![feature(min_const_generics)]
+#![no_std]
+#![deny(missing_docs)]
```

Comment:
```suggestion
#![feature(min_const_generics)]
#![no_std]
#![forbid(unsafe_code)]
#![deny(missing_docs)]
```

Reply:
once rebased, `min_const_generics` is no longer needed